### ISSUE TITLE
Bare Workflow: Updates.reloadAsync() makes my android app crash

### ISSUE THREAD

**Issue by axelmarciano** [expo/expo#7737]
Hey guys,
I used to work on ExpoKit and i've just switch to Bare Workflow with the new release (SDK 37).
Everything is working fine except for the OTA update.
I'm using the manual update for my application by calling the checkForUpdateAsync and the fetchUpdateAsync methods.

Those two methods are working (the update is fetched) but when i'm trying to reload my app after this with reloadAsync method my android app crashed with a strange error (on release).

Yet, when I'm restarting the application, the update is available.

This log is extracted from adb logcat --buffer=crash:

`04-08 14:30:35.158 29276 29598 E AndroidRuntime: FATAL EXCEPTION: create_react_context
04-08 14:30:35.158 29276 29598 E AndroidRuntime: Process: XXXXX, PID: 29276
04-08 14:30:35.158 29276 29598 E AndroidRuntime: com.facebook.jni.CppException: facebook::react::Recoverable: Could not open file/data/user/0/XXXX/files/.expo-internal/7C6DA04432421D9566434203858FEC7C132FA97C43BD314556629A1673F750F6.js: No such file or directory
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at com.facebook.react.bridge.CatalystInstanceImpl.jniLoadScriptFromFile(Native Method)
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at com.facebook.react.bridge.CatalystInstanceImpl.loadScriptFromFile(Unknown Source:2)
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at com.facebook.react.bridge.JSBundleLoader$2.loadScript(Unknown Source:6)
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at com.facebook.react.bridge.CatalystInstanceImpl.runJSBundle(Unknown Source:18)
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at com.facebook.react.s.a(Unknown Source:154)
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at com.facebook.react.s.a(Unknown Source:0)
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at com.facebook.react.s$e.run(Unknown Source:68)
04-08 14:30:35.158 29276 29598 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:764)
`

**Comment by cruzach** [comment]
Hey @axelmarciano sorry you're having trouble with this. Could you share a sample of the code you're using?

**Comment by axelmarciano** [comment]
Yes ofc.
I'm using an hook useUpdater(). This hook is called on an <Updater/> Component with a simple logic :
- When the app is fetching a new update, an opacity loader is shown on the screen
- A interval is launched with the component to check for update every X minutes.

`
/* global __DEV__*/
import { useState } from 'react';
import * as Updates from 'expo-updates';
import { Alert } from 'react-native';
import { t } from '../config/i18n';

export const useUpdater = () => {
  const [loading, load] = useState(false);

  const reloadAppFromCache = () => {
    Updates.reloadAsync();
  };

  const updateChecker = async (withAlert) => {
    if (__DEV__) {
      return;
    }
    const update = await Updates.checkForUpdateAsync();
    if (update.isAvailable) {
      load(true);
      await Updates.fetchUpdateAsync();
      // TO-DO ... notify user of update ...
      load(false);
      reloadAppFromCache();
    } else if (withAlert) {
      Alert.alert(
        t('restaurateur:update:notAvailable:title'),
        t('restaurateur:update:notAvailable:body'),
        [
          {
            text: t('restaurateur:update:notAvailable:button'),
            style: 'cancel',
          },
        ],
        { cancelable: false }
      );
    }
  };

  return {
    loading,
    updateChecker,
    reloadAppFromCache,
  };
};

`

**Comment by cruzach** [comment]
Hm, do you think you could share a blank project that reproduces the same crash, not just the `useUpdater` hook? 

Also, are you checking for an update every few minutes?

**Comment by changLiuUNSW** [comment]
Have the same issue. reloadAsync make app crash

**Comment by brentvatne** [comment]
> Have the same issue. reloadAsync make app crash

please share a reproducible example. https://stackoverflow.com/help/minimal-reproducible-example

**Comment by hoangtrucit** [comment]
i have sample problem and this is log when i use method `reloadAsync`
> Fatal Exception: com.facebook.jni.CppException: facebook::react::Recoverable: Could not open file/data/user/0/com.production.nhiemvu/files/.expo-internal/89FDF764C4FB0B609B9C64D83335FD5BE56A655BCD9FC8EE45500D9A602EA567.js: No such file or directory
       at com.facebook.react.bridge.CatalystInstanceImpl.jniLoadScriptFromFile()
       at com.facebook.react.bridge.CatalystInstanceImpl.loadScriptFromFile(:2)
       at com.facebook.react.bridge.JSBundleLoader$2.loadScript(:6)
       at com.facebook.react.bridge.CatalystInstanceImpl.runJSBundle(:18)
       at com.facebook.react.M.a(:154)
       at com.facebook.react.M.a()
       at com.facebook.react.H.run(:68)
       at java.lang.Thread.run(Thread.java:764)

**Comment by YahiaBadr** [comment]
same for me happens in android 8.1

**Comment by YahiaBadr** [comment]
I did a new repo and i tried the same code i have for just the updating process, however, the app didn't crash which was unexpected. So now it means they our code could contain something wrong or not in the right place. I will try to figure it out with try and error and hopefully, i will find the problem.

**Comment by YahiaBadr** [comment]
I found my mistake as i expected !!!
I was calling the updating function inside the component called `<RootStack/>`, but when i change the place of the calling to be in the App.js outside any other component it worked.
```
 export default class App extends React.Component {
  state={
    updating:0
  }
  componentDidMount(){
    checkForUpdates.bind(this)();
  }
  render(){
    if(this.state.updating !==0) return(
      <Container style={{ backgroundColor: theme, display:"flex" }}>
				<Loading disabled={this.state.updating===2} updating />
      </Container>
    )
  return (
    <Root>
      <Provider store={store}>
        <SafeAreaProvider>
          <SafeAreaView
            style={{ flex: 1, backgroundColor: theme2, paddingTop:0}} 
            forceInset={{ top: 'never', bottom:'always' }}
          >
            <NavigationContainer>
              <RootStack/>
            </NavigationContainer>
          </SafeAreaView>
        </SafeAreaProvider>
      </Provider>
    </Root>
  );
}
}
```

**Comment by baronchng** [comment]
Hey @axelmarciano, I just recently faced and solved this issue. From your code, I think it's better you do `await Updates.reloadAsync();` so that you have control of what happens when the `Promise` is resolved. 

But most importantly, you shouldn't run any code after `Updates.reloadAsync()` such as `setState()` despite the documentation said it's **not recommended** when actually it seems like we **cannot**.  Refer to `Updates.reloadAsync()` section in [Expo Updates Doc](https://docs.expo.io/versions/latest/sdk/updates/). 

Another thing to watch out is if `reloadAsync()` is used in `App()` component where normally there are a few `useEffect()` around, ensure that no other `useEffect()` will actually do something after you call `reloadAsync()`.

What happened to me was 
```
try {
  ... check for updates and fetch

  await Updates.reloadAsync(); <---- we shouldn't do anything after this. 
}
catch (error) {
}
finally {
  setState(); <---- this happens right after the reloadAsync() above and crashes the Android app
}
```

So I changed to something like
```
try {
  const update = await Updates.checkForUpdateAsync();
  if (update.isAvailable) {
    await Updates.reloadAsync(); 

    // don't do anything
  } else {
    setState(); <---- since the reload above will make update.isAvailable to false, this will run
  }
}
catch (error) {
}
```

**Comment by jonathanstiansen** [comment]
I'll work on a repo for this as well, and see if we can get you guys something @brentvatne we are getting one per android update right now, we call reload after loading the update. It seems like something with the new Update module because we've had this code around for a while.

**Comment by jonathanstiansen** [comment]
We've been having tons of trouble around this code that only started since we updated from 2.7 to Updates >2.9, we were on 2.9 then updated to 2.11 to see if this would be fixed. No such luck so just chipping in to say "me too" but I'll actually try to get the repo done.

**Comment by esamelson** [comment]
Hi all - I've had no luck reproducing this issue using the fragments provided above. As it's been months since there's been any activity on this thread, I'm going to go ahead and close it. If you continue to run into this crash and can provide a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) as was requested earlier in the thread, please open a new issue. Thanks 🙂

**Comment by sfalihi** [comment]
> Hey @axelmarciano, I just recently faced and solved this issue. From your code, I think it's better you do `await Updates.reloadAsync();` so that you have control of what happens when the `Promise` is resolved. I don't know where you use the hook but i will just make assumptions.
> 
> Most importantly, you shouldn't run any code after `Updates.reloadAsync()` such as `setState()` despite the documentation said it's **not recommended** when actually it seems like we **cannot**. Refer to `Updates.reloadAsync()` section in [Expo Updates Doc](https://docs.expo.io/versions/latest/sdk/updates/).
> 
> Another thing to watch out is if `reloadAsync()` is used in `App()` component where normally there are a few `useEffect()` around, ensure that no other `useEffect()` will actually do something after you call `reloadAsync()`.
> 
> What happened to me was
> 
> ```
> try {
>   ... check for updates and fetch
> 
>   await Updates.reloadAsync(); <---- we shouldn't do anything after this. 
> }
> catch (error) {
> }
> finally {
>   setState(); <---- this happens right after the reloadAsync() above and crashes the Android app
> }
> ```
> 
> So I changed to something like
> 
> ```
> try {
>   const update = await Updates.checkForUpdateAsync();
>   if (update.isAvailable) {
>     await Updates.reloadAsync(); 
> 
>     // don't do anything
>   } else {
>     setState(); <---- since the reload above will make update.isAvailable to false, this will run
>   }
> }
> catch (error) {
> }
> ```

i had same issue when i was trying to reloadAsync for changing app direction at start. (ltr - rtl)
i solved it by making sure that **no code runs or no rendering happens after reloadAsync** as suggested before. tnx

**Comment by hpeinar** [comment]
Adding to this thread as we ran into the same problem on Android but with different kind of error message. This thread gave us hints what to try to fix this.

We're using:
`expo 51.0.5`
`expo-updates 0.25.11`
`react-native 0.74.1`

**The fix:** As mentioned above, we made sure the application didn't use any kind of `useState` logic after `reloadAsync()` had been called. We edited our application flow in a way that the main `<App />` component doesn't even get inserted when an OTA updated was found. There were some other minor changes needed to make sure the `reloadAsync()` call is indeed last one in the initialization flow.

We tested the fix on preview channel where we could easily test OTA releasing and crashing.

Released the fix with OTA and our Android application no longer crashing when applying OTA update.

Attaching both screenshot of how the error looks like in Android Studio and text version so search engines would pick it up.

```Abort message: 'JNI DETECTED ERROR IN APPLICATION: obj == null in call to GetObjectField
from void com.facebook.jni.NativeRunnable.run()'```

<img width="1660" alt="Screenshot 2024-10-01 at 16 48 13" src="https://github.com/user-attachments/assets/d8694193-32f3-4911-8892-b510dc23a57e">
